firebase_cloud_pkg/functions/src/index.ts

```diff
@@ -0,0 +1,112 @@
+import * as functions from 'firebase-functions';
+import * as admin from 'firebase-admin';
+import * as apptreeio from 'apptreeio'
+import {GetDocumentInput, JsonObject, QueryAndQueueInput, QueryInput, QueryParameters, UpsertInput} from "./models";
+
+admin.initializeApp();
+
+export const getDocument = functions.https.onRequest(async (req, resp) => {
+    const inputs = req.body as GetDocumentInput;
+    if (!inputs.RecordPath || inputs.RecordPath == "") {
+        resp.status(500).send('record path is a required input');
+        return;
+    }
+
+    try {
+        const docRef = admin.firestore().doc(inputs.RecordPath);
+        const snapshot = await docRef.get();
+        if (!snapshot.exists) {
+            resp.send({Success: false, Message: "not found"});
+        } else {
+            resp.send({
+                Success: true,
+                Record: snapshot.data() as JsonObject,
+            });
+        }
+    } catch (e) {
+        resp.status(500).send(`error encountered: ${e}`);
+    }
+});
+
+export const query = functions.https.onRequest(async (req, resp) => {
+    const inputs = req.body as QueryInput;
+    if (!inputs.CollectionPath || inputs.CollectionPath === "") {
+        resp.status(500).send(`collection path is a required input`);
+        return;
+    }
+
+    try {
+        const records = await queryCollection(inputs.CollectionPath, inputs.QueryParams);
+        resp.send({"Records": records});
+    } catch (e) {
+        resp.status(500).send(`error encountered: ${e}`);
+    }
+});
+
+export const queryAndQueue = functions.https.onRequest(async (req, resp) => {
+    const client = apptreeio.createClient(req);
+
+    const inputs = req.body as QueryAndQueueInput;
+    if (!inputs.CollectionPath || inputs.CollectionPath === "") {
```

Comment:
There is a method I added `apptreeio.validateInputs`. That might be a cleaner way to do this.